### PR TITLE
Model parallel code changes + context parallelism

### PR DESCRIPTION
**Pull request by ShashankSubramanian**
* Changed communicator wireups to use RankGenerator (and associated comm names) from Megatron 
* Removed comm hooks for shared weights for TP
* Added orthogonal context parallelism (CP) and retained comm hooks for shared weights in CP group
* Added unit tests for distributed layers

### CODE REVIEW

File: distributed/helpers.py [hunk 1]
```diff
@@ -1,23 +1,20 @@
...
         for param in model.parameters():
             if not hasattr(param, "is_shared_mp"):
-                param.is_shared_mp = ["model"]
+                # all sharded weights manually annotate this field
+                # if weight doesnt have annotation, then it is a shared weight
+                # layers like patch-embed, decoder head, pos-embed are fully
+                # shared (and not sharded) in this example
+                param.is_shared_mp = ["tp-cp"] # only TP-CP implemented for now
+                # careful about this stuff..
+                param.mark_for_reduction = [] # not all params need special handling
```
Comment (azrael417) [inline distributed/helpers.py:17]: I think the mark for reduction is not necessary, why not use is_shared_mp for that purpose. What we have in Makani is the following: we annotate each tensor what its sharing is, f the tensor has no is_shared_mp then we default to some behavior (In this case not do anything special with it). Parameters is per module parameter, so you can have both inside the same module, it is no problem.

File: distributed/helpers.py [hunk 2]
```diff
@@ -31,85 +28,125 @@ def sync_params(model):
...
-    if use_fp32:
-        dtype = input_.dtype
-        inputf_ = input_.float().contiguous()
-        dist.all_reduce(inputf_, group=group)
-        input_ = inputf_.to(dtype)
-    else:
-        input_ = input_.contiguous()
-        dist.all_reduce(input_, group=group)
+    input_ = input_.contiguous()
+    dist.all_reduce(input_.contiguous(), group=comm.get_group(comm_name))
```
Comment (azrael417) [inline distributed/helpers.py:60]: The second .contiguous can be dropped here, since input_ is already made config before

File: distributed/layers.py [hunk 3]
```diff
@@ -43,20 +61,15 @@ def __init__(
...
-        ]
+        self.weight.is_shared_mp = [comm_act_name] # weights are sharded in tp but shared across cp
+        self.weight.mark_for_reduction = [comm_act_name] # shared weights must be additionally reduced
         if bias:
             self.bias = nn.Parameter(torch.ones(1, 1, out_dim_local))
-            self.bias.is_shared_mp = comm_names_shared
-            self.bias.sharded_dims_mp = [None, self.comm_out_name, None, None]
+            # if inp dim of W is sharded, then the bias is shared across this group and also
+            # shared in cp grp
+            self.bias.is_shared_mp = [self.comm_inp_name, comm_act_name]
```
Comment (azrael417) [inline distributed/layers.py:71]: The bias should only be reduced over comm_act_name, so we just can add this to the is_shared_mp variable. We do not need the second one I think, or are we using that in a different context? For initial parameter broadcasting?

File: distributed/mappings.py [hunk 4]
```diff
@@ -128,36 +193,143 @@ def init_ddp_model_and_reduction_hooks(model,
...
+                # divide by dp size since we need to average the grads
+                coalesced.div_(dp_size)
+                # Check if coalesced contains NaNs or Infs after division
+                if torch.isnan(coalesced).any() or torch.isinf(coalesced).any():
+                    raise ValueError(f"NaN or Inf detected in coalesced gradients for group: {combined_group}")
+
+                group = comm.get_group(combined_group)
+                # all-reduce in the appropriate combined group
+                fut_group = dist.all_reduce(coalesced, op=dist.ReduceOp.SUM, group=group, async_op=True).get_future()
+                # copy the results to the gradients (update the grads basically)
```
Comment (azrael417) [inline distributed/mappings.py:321]: This is the issue I was mentioning on the last call: the future is marked done when the reduction is enqueued, which in turn means that the grad buffer copy can fire off too early, leading to data corruption. Therefore, as in my original routine above, I recommend marking this one blocking (so not async) and then follow it up by a blocking copy. The performance impact is not very big I think.

The whole future business is not very well documented unfortunately and it took us quite a while to debug this logic. In my above logic, I was also using composite groups implicitly. For example if I have an image h,w decomposed, then I also have a "spatial" group which is the union of h and w. And in case weights are shared across h and w and need to be reduced, I am just using the parent group. So basically I decompose the communicator into a graph, with the most fine grained comms in the leaves.

**Review by azrael417 — COMMENTED** [review]
Thank you very much Shashank, I think the MR looks good. I think what would make some sense to check the grads is to use some extended formalism like we have in torch harmonics:

https://github.com/NVIDIA/torch-harmonics/blob/main/tests/test_distributed_sht.py

The idea is: I am creating a non distributed layer and then a distributed layer of the same type with exactly the same wights (I basically distribute or split the weights accordingly). Then I am running forward and backward pass und dann vergleiche ich input grads von backward und forward outputs. We could do that for our layers and additionally maybe check the weight grads as well in the same manner. We do not need to implement that for every layer but maybe for the attention one. It would help to understand where the issues are. Feel free to use the TH distributed test as template for implementing that, it has a lot of the gathering/distribution/init logic already implemented. Please let me know what you think.

File: distributed/helpers.py (hunk 1 above)
Comment (ShashankSubramanian) [reply distributed/helpers.py:17]: I wanted to distinguish between shared params (shared across some model parallel group) which need the same init, and shared params that need an additional allreduce due to the input activation gradient being sharded across the same model parallel group. These two sets of parameters are unequal I think (some shared params do not need the additional allreduce; example, if you look at the decoder head, a linear layer, it shares it parameters across the full model parallel group but the input activation gradient is also shared - so the wgrads will be in sync. The only allreduce they need is across the data parallel group where the input activation gradient is also shared (each model parallel group gets a different batch) and that allreduce is the normal DDP allreduce (but averaged due to the MSE loss). Do you feel this is correct?

File: distributed/layers.py (hunk 3 above)
Comment (ShashankSubramanian) [reply distributed/layers.py:71]: Yeah for init param syncs (see previous comment)

File: distributed/mappings.py (hunk 4 above)
Comment (ShashankSubramanian) [reply distributed/mappings.py:321]: Ok got it. I spent some time in trying to understand the docs and thought the `then()` logic (https://pytorch.org/docs/stable/futures.html#torch.futures.Future.then) would ensure that the grad buffer copy would start only after the reduction is complete. So, the async reduction is enqueued, the future returns and the _sync_gradients is chained, but any operation on the gpu would start only after the prev op (the reduction) finishes. I could be very wrong.. (and think its painful to debug) and happy to revert to the blocking logic to be safe.

Comment (azrael417) [reply distributed/mappings.py:321]: Hmm, this might work. Have you tested this at larger scales and does it run?

File: distributed/helpers.py (hunk 1 above)
Comment (azrael417) [reply distributed/helpers.py:17]: I think it is right

Comment (azrael417) [reply distributed/helpers.py:17]: The question is, is there a simple way to identify whether the grads need reductions or not? I guess not without some user intervention, or? That would be good to think about.

File: distributed/mappings.py (hunk 4 above)
Comment (azrael417) [reply distributed/mappings.py:321]: Please have a look here though: https://pytorch.org/docs/stable/_modules/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.html#PowerSGDState

In this more complicated hook, they do not simply return the future from the async AR but also wait for it to complete. I think something like that needs to be done, but I am really not sure. That logic is very confusing.
